src/skin/qml/asyncimageprovider.cpp

```diff
@@ -0,0 +1,85 @@
+#include "skin/qml/asyncimageprovider.h"
+
+#include "library/coverartcache.h"
+#include "track/track.h"
+
+namespace {
+const QString kCoverArtPrefix = QStringLiteral("coverart/");
+}
+
+namespace mixxx {
+namespace skin {
+namespace qml {
+
+AsyncImageResponse::AsyncImageResponse(
+        std::shared_ptr<TrackCollectionManager> pTrackCollectionManager,
+        const QString& id,
+        const QSize& requestedSize)
+        : m_pTrackCollectionManager(pTrackCollectionManager),
+          m_id(id),
+          m_requestedSize(requestedSize) {
+    setAutoDelete(false);
+}
+
+QQuickTextureFactory* AsyncImageResponse::textureFactory() const {
+    return QQuickTextureFactory::textureFactoryForImage(m_image);
+}
+
+void AsyncImageResponse::run() {
+    if (m_id.startsWith(kCoverArtPrefix)) {
+        QString trackLocation = AsyncImageProvider::coverArtUrlIdToTrackLocation(m_id);
+
+        TrackPointer pTrack = m_pTrackCollectionManager->getTrackByRef(
```

Comment:
Ideally all I/O should happen in a separate thread. But that's not possible with our library.

Temporary workaround:
1. Load the track in the constructor synchronously
2. Only perform the loading and scaling of the image asynchronously in run()

Reply:
No, that doesn't work either. The constructor is called in the QML renderer thread, not the  regular GUI thread I think.